keps/sig-api-machinery/5073-declarative-validation-with-validation-gen/README.md

```diff
@@ -944,23 +948,82 @@ The linter, as previously described, will enforce rules to address valid zero-va
+3.  **Conditional Validation:** Validation rules can use a special `subresource` parameter (e.g., `subresource == '/status'`) to apply conditional logic. This allows rules to behave differently depending on whether the update comes via the primary resource or a specific subresource.
+4.  **Ratcheting:** Declarative Validation uses ratcheting, meaning validation does not fail for fields that have not changed from the existing stored object. Combined with field wiping, this scopes validation to only the subset of fields that the subresource operation is intended and permitted to modify.
+
+**Validation Examples:**
+
+* An update via `pods/status` first has its `spec` field changes wiped by the Pod strategy. Then, the entire Pod object is validated using the standard Pod validation rules. Ratcheting skips checks on unchanged `metadata` or `status` fields.
+* An update via `pods/resize` is validated using the standard Pod rules. However, a rule on `spec.container[*].resources` might look like `+k8s:exceptSubresource("/resize"')=+k8s:immutable`, effectively enforcing immutability *unless* the update comes via the `resize` subresource.
+
+**Support required:**
+
+* Conditional validation, provided by dedicated `+k8s:subresource` and `+k8s:exceptSubresource` tags and a `subresources` parameter within validation rule expressions (e.g. `+k8s:if('subresource != "/resize"')=+k8s:immutable`).
+
+#### Scale-Type Subresources
+
+These subresources have the following characteristics:
+
+* They often represent a different API `kind` (e.g., `autoscaling/v1.Scale`) than the resource they modify (e.g., `apps/v1.Deployment`).
+* Despite being a different `kind`, updates to the subresource modify fields within the underlying storage object (same object in etcd) as the *primary* resource.
+
+**Examples:**
+
+* An update to `deployments/scale` modifies the `spec.replicas` field within the primary `deployments` resource.
+* Creating a `Binding` object via `pods/binding` sets the `spec.nodeName` field on the primary `pods` resource.
+
+**Validation Process:**
+
+1.  **Subresource Validation:** The incoming subresource object itself (e.g., the `autoscaling/v1.Scale` object) is validated using *its own* declarative rules.
+2.  **Storage Layer Application:** The resource's storage layer logic translates the validated subresource update into changes on the primary resource's fields (e.g., mapping the `Scale` object's `spec.replicas` to the `Deployment` object's `spec.replicas`).
+3.  **Primary Resource Validation:** The *modified primary resource* (e.g., `Deployment`) is then validated using *its* standard declarative validation rules.
+4.  **Ratcheting:** Ratcheting is applied during the primary resource validation, skipping checks on fields that were not affected by the subresource update.
```

Comment:
Do we ratchet on the `autoscaling/v1.Scale` object?  I realized in reading this - I have no idea what API operation is presented when accessing /scale -- do we support POST?  Or just PUT and other updates?  Do we have the "old" available for validation?

Reply:
> Do we ratchet on the autoscaling/v1.Scale object?

I think we should support it.  I don't believe we have any ratcheting on scale today. (We do have status ratcheting...)

> do we support POST? Or just PUT and other updates?

Just PUT, PATCH:
```
jq '.paths."/api/v1/namespaces/{namespace}/replicationcontrollers/{name}/scale" | keys' api/openapi-spec/v3/api__v1_openapi.json
[
  "get",
  "parameters",
  "patch",
  "put"
]
```

> Do we have the "old" available for validation?

Yes, we always have it. ([storage rest uses UpdatedObjectInfo](https://github.com/kubernetes/kubernetes/blob/5d4b4a160dc551dc8979012eeabea1a098945603/staging/src/k8s.io/apiserver/pkg/registry/rest/rest.go#L240))

Reply:
I'll summarize in KEP for posterity.